packages/tpetra/core/ext/TpetraExt_TripleMatrixMultiply_def.hpp

```diff
@@ -467,19 +469,62 @@ namespace Tpetra {
 
       // Call the actual kernel.  We'll rely on partial template specialization to call the correct one ---
       // Either the straight-up Tpetra code (SerialNode) or the KokkosKernels one (other NGP node types)
-      KernelWrappers3MMM<Scalar,LocalOrdinal,GlobalOrdinal,Node>::mult_R_A_P_newmatrix_kernel_wrapper(Rview,
-                                                                                                      Aview,
-                                                                                                      Pview,
-                                                                                                      Acol2PRow,
-                                                                                                      Acol2PRowImport,
-                                                                                                      Pcol2Accol,
-                                                                                                      PIcol2Accol,
-                                                                                                      Ac,
-                                                                                                      Acimport,
-                                                                                                      label,
-                                                                                                      params);
+      KernelWrappers3MMM_Specialized<Scalar,LocalOrdinal,GlobalOrdinal,Node>::
+        mult_R_A_P_newmatrix_kernel_wrapper(Rview, Aview, Pview,
+                                            Acol2PRow, Acol2PRowImport, Pcol2Accol, PIcol2Accol,
+                                            Ac, Acimport, label, params);
     }
 
+    /*********************************************************************************************************/
+    template<class InRowptrArrayType, class InColindArrayType, class InValsArrayType,
+             class OutRowptrType, class OutColindType, class OutValsType>
+    void copy_out_from_thread_memory(const InRowptrArrayType & Inrowptr, const InColindArrayType &Incolind, const InValsArrayType & Invalues,
+                                       size_t m, double thread_chunk,
+                                       OutRowptrType & row_mapC, OutColindType &entriesC, OutValsType & valuesC ) {
+      typedef OutRowptrType lno_view_t;
+      typedef OutColindType lno_nnz_view_t;
+      typedef OutValsType scalar_view_t;
+      typedef typename lno_view_t::execution_space execution_space;
+      typedef Kokkos::RangePolicy<execution_space, size_t> range_type;
+
+      // Generate the starting nnz number per thread
+      size_t thread_max =  Inrowptr.size();
+      size_t c_nnz_size=0;
+      lno_view_t thread_start_nnz("thread_nnz",thread_max+1);
+      Kokkos::parallel_scan("LTG::Scan",range_type(0,thread_max).set_chunk_size(1), [=] (const size_t i, size_t& update, const bool final) {
```

Comment:
Tpetra already has a function for turning counts into offsets: https://github.com/trilinos/Trilinos/blob/1a9e0546b639dad0abff827cd825c45e301d46af/packages/tpetra/core/src/Tpetra_Details_computeOffsets.hpp#L284

Reply:
@mhoemmen I think since this particular scan is summing the lengths in a view-of-views, computeOffsetsFromCounts can't be used. I could first make another View<size_t*> to hold the lengths. I'm not sure which one saves on complexity.

Reply:
@mhoemmen Actually I will make the separate View, this avoids having a whole extra lambda.